src/fsharp/CompileOps.fs

```diff
@@ -755,23 +750,52 @@ let OutputPhasedErrorR (os: StringBuilder) (err: PhasedDiagnostic) (suggestNames
           os.Append(e.ContextualErrorMessage) |> ignore
 #endif
 
-      | UnresolvedOverloading(_, _, mtext, _) -> 
-          os.Append mtext |> ignore
+      | UnresolvedOverloading(denv, callerArgs, failure, m) ->
+          let nl = System.Environment.NewLine
+          let displayArgType (name, ttype) =
+              let typeDisplay = NicePrint.prettyStringOfTy denv ttype
```

Comment:
The technique to display types needs to be adjusted.  The rule is "format all the types in one go".  This ensures that type inference variables that are actually the same across the different types get formatted to the same text.  See `minimalStringsOfTwoTypes` for example for how we format two types (which in that case are then going to be compared).  There are also helpers to format multiple types I believe.

For test cases, you should I think notice the difference when types cross arguments are linked, e.g. `C.M(x,x)` where each `x` has an unknown inference type, or `C.M(x,x,y,y)`

Reply:
> See `minimalStringsOfTwoTypes` for example for how we format two types (which in that case are then going to be compared). There are also helpers to format multiple types I believe.

I'm going to look further into this, I've just hit https://github.com/microsoft/visualfsharp/issues/6725.

> For test cases, you should I think notice the difference when types cross arguments are linked, e.g. C.M(x,x) where each x has an unknown inference type, or C.M(x,x,y,y)

Could you tell if:

```fsharp
type C() =
    static member M(a, b) = a + b
    static member M(a, b, c) = a + b + c
    static member M(a, b, c, d, e) = float (a + b + c + d) * e
    static member M(a, b, c, d, e) = decimal (a + b + c + d) * e
    static member M(a, b, c, d, e) = float32 (a + b + c + d) * e
C.M(1, 2, 3, 4, 5)
```

is a relevant case for what you described?

it currently gives:

![image](https://user-images.githubusercontent.com/87944/57675308-73565680-75d6-11e9-8f91-cd5b0c76a0c3.png)

and current state of this PR:

![image](https://user-images.githubusercontent.com/87944/57675604-4b1b2780-75d7-11e9-90a7-a4f9f4e00afc.png)

When generics are involved, I'm not sure the changes make it more readable, do we want to do something special for those cases?

Reply:
@dsyme:

> The technique to display types needs to be adjusted. The rule is "format all the types in one go". This ensures that type inference variables that are actually the same across the different types get formatted to the same text.

Looking at original code, I don't see that rule being considered, also, does it make sense to consolidate generic arguments names (`'a`, `'b`, etc.)  over a list of many overloads in the error message?

It seems to me it would lead to messages with generic arguments named `'z` without having 26 type parameters on a given overload.

Please see in original code, the stringification happens in a List.map and doesn't seem to pass any state that would be mutated (or I've overlooked that):

https://github.com/dotnet/fsharp/pull/6596/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsi&file-filters%5B%5D=.fsx#diff-641da7c52ffc888699888ebd03559cb5L175

https://github.com/dotnet/fsharp/pull/6596/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsi&file-filters%5B%5D=.fsx#diff-641da7c52ffc888699888ebd03559cb5L2552

So in the context of type id "unification/prettification", I don't see this PR making any change of behaviour compared to existing code, am I still overlooking something on that matter?

Reply:
Can you prettify the group of arguments for a single overload in a single prettification? (Not all the overloads at once)

Reply:
@dsyme, right makes more sense than what I said wouldn't work :slightly_smiling_face:.

AFAIU, this is handled by call to `stringOfMethInfo`, call which has remained in this PR.

It ends up there: 
https://github.com/dotnet/fsharp/blob/ccb913d3a05863e5b1861d64994ffb97ed498855/src/fsharp/NicePrint.fs#L1381-L1397

Reply:
I found a case where the "Known types of arguments" would hit the issue @dsyme is concerned about:

```fsharp

type A =
  static member Foo(a: 'a, b: 'a -> 'b, c: 'a -> 'b, d: 'a -> 'b) : 'b = b a
  static member Foo(a: 'a, b: 'a -> 'b, c: 'b -> 'c, d: 'c -> 'd) : 'd = d (c( b a))

let inline f a b c =
  A.Foo(1., a, b, c)
```

printing

>  A unique overload for method 'Foo' could not be determined based on type information prior to this program point. A type annotation may be needed.
>
> Known types of arguments: __float * 'a * 'a * 'a__
>
>Candidates:
> - static member A.Foo : a:'a * b:('a -> 'b) * c:('a -> 'b) * d:('a -> 'b) -> 'b
> - static member A.Foo : a:'a * b:('a -> 'b) * c:('b -> 'c) * d:('c -> 'd) -> 'd

I gather we should have __float * 'a * 'b * 'c__ here.

Going to look more at NicePrint.fs and `Prettify*` functions in tastops.fs.